opentelemetry-api/src/opentelemetry/metrics/instrument.py

```diff
@@ -0,0 +1,208 @@
+                )
+                self._callback = callback
+
+    @property
+    def callback(self):
+        def function():
+            measurement = next(self._callback)
+            if not isinstance(measurement, Measurement):
+                _logger.error("Callback must return a Measurement")
+                return None
+
+            return measurement
+
+        return function
+
+
+class _Adding(Instrument):
+    pass
+
+
+class _Grouping(Instrument):
+    pass
+
+
+class _Monotonic(_Adding):
+    pass
+
+
+class _NonMonotonic(_Adding):
+    pass
+
+
+class Counter(_Monotonic, Synchronous):
+    @abstractmethod
+    def add(self, amount, attributes=None):
+        with self._lock:
+            if amount < 0:
+                _logger.error("Amount must be non-negative")
+
+
+class DefaultCounter(Counter):
+    def __init__(self, name, unit="", description=""):
+        super().__init__(name, unit=unit, description=description)
+
+    def add(self, amount, attributes=None):
+        with self._lock:
+            return super().add(amount, attributes=attributes)
+
+
+class UpDownCounter(_NonMonotonic, Synchronous):
+    @abstractmethod
+    def add(self, amount, attributes=None):
+        with self._lock:
+            pass
+
+
+class DefaultUpDownCounter(UpDownCounter):
+    def __init__(self, name, unit="", description=""):
+        super().__init__(name, unit=unit, description=description)
+
+    def add(self, amount, attributes=None):
+        with self._lock:
+            return super().add(amount, attributes=attributes)
+
+
+class ObservableCounter(_Monotonic, Asynchronous):
+    @property
+    def callback(self):
+        def function():
+            measurement = super(ObservableCounter, self).callback()
+
+            if measurement is not None and measurement.value < 0:
+                _logger.error("Amount must be non-negative")
+                return None
+
+            return measurement
+
+        return function
+
+
+class DefaultObservableCounter(ObservableCounter):
+    def __init__(self, name, callback, unit="", description=""):
+        super().__init__(name, callback, unit=unit, description=description)
+
+
+class ObservableUpDownCounter(_NonMonotonic, Asynchronous):
+
+    pass
+
+
+class DefaultObservableUpDownCounter(ObservableUpDownCounter):
+    def __init__(self, name, callback, unit="", description=""):
+        super().__init__(name, callback, unit=unit, description=description)
+
+
+class Histogram(_Grouping, Synchronous):
+    @abstractmethod
+    def record(self, amount, attributes=None):
+        with self._lock:
+            pass
+
+
+class DefaultHistogram(Histogram):
+    def __init__(self, name, unit="", description=""):
+        super().__init__(name, unit=unit, description=description)
+
+    def record(self, amount, attributes=None):
+        with self._lock:
+            return super().record(amount, attributes=attributes)
```

Comment:
Aren't these supposed to be in SDK?

Reply:
What are you referring to with "these"?

Reply:
Nevermind, I confused it with implementation part.